proposals/module-linking/Binary.md

```diff
@@ -0,0 +1,219 @@
+
+instancedef ::= m:moduleidx e*:vec(exportdesc)        ->    {instantiate m, imports e}
+```
+
+This defines instances in the module, appending to the instance index space.
+Each instance definition declares the module it's instantiating as well as the
+items used to instantiate that instance.
+
+**Validation**
+
+* The type index `m` must point to a module type.
+* The length of `e*` must be the same as the number of imports the module type
+  has
+* The type of each element of `e*` must be a subtype of the type that it's being
+  matched with. Matching happens pairwise with the list of imports on the module
+  type for `m`.
+* Indices of items referred to by `exportdesc` are all in-bounds. Can only refer
+  to imports/previous aliases, since only those sections can precede.
+
+## Alias section (102)
+
+A new module section is added
+
+```
+aliassec ::=  a*:section_102(vec(alias))     ->        a*
+
+alias ::=
+    0x00 i:instanceidx 0x00 e:exportidx      ->       (alias (instance $i) (func $e))
+    0x00 i:instanceidx 0x01 e:exportidx      ->       (alias (instance $i) (table $e))
+    0x00 i:instanceidx 0x02 e:exportidx      ->       (alias (instance $i) (memory $e))
+    0x00 i:instanceidx 0x03 e:exportidx      ->       (alias (instance $i) (global $e))
+    0x00 i:instanceidx 0x05 e:exportidx      ->       (alias (instance $i) (module $e))
+    0x00 i:instanceidx 0x06 e:exportidx      ->       (alias (instance $i) (instance $e))
+    0x01 0x04 m:moduleidx                    ->       (alias parent (module $m))
+    0x01 0x07 t:typeidx                      ->       (alias parent (type $t))
+```
+
+**Validation**
+
+* Aliased instance indexes are all in bounds
+* Aliased instance export indices are in bounds relative to the instance
+* Export indices match the actual type of the export
+* Aliases append to the respective index space
+
+## Function section
+
+**Validation**
+
+* Type indices must point to function types.
+
+## Export section
+
+update
+[`exportdesc`](https://webassembly.github.io/spec/core/binary/modules.html#binary-exportdesc)
+
+```
+exportdesc ::=
+    ...
+    0x05 x:moduleidx                        -> module x
+    0x06 x:instanceidx                      -> instance x
+```
+
+**Validation**
+
+* Module/instance indexes must be in-bounds.
+
+## Module Code Section (103)
+
+```
+modulecodesec ::= m*:section_103(vec(modulecode))       ->    m*
+
+modulecode ::= size:u32 mod:module                      ->    mod, size = ||mod||
+```
+
+Note that this is intentionally a recursive production where `module` refers to
+the top-level
+[`module`](https://webassembly.github.io/spec/core/binary/modules.html#binary-module)
+
+**Validation**
+
+* Module definitions must match their module type exactly, no subtyping (or
+  maybe subtyping, see WebAssembly/module-linking#9).
+* Modules themselves validate recursively.
+* Must have the same number of modules as the count of all local module
+  sections.
```

Comment:
Oh, one more thing: It would be good to add a bullet here speaking to how `parent` aliases are validated in nested modules.  In particular, I think it's necessary to pass in the subset of the module and type index spaces (that existed at the point of the module declaration in the `module` section) into the validation rules of the nested module (as a form of context).

Reply:
Ah yes good to include!

To clarify, though, there's no restriction on what a submodule can alias, right? It can reference any type and any module, no matter where the type or module is defined in the parent? (relative to this module's own definition in the `module` section)

Reply:
I think once we get type exports, if a nested module can access the entire type index space, you could end up with nasty cross-module type cycles.  Thus, I had been thinking that a module definition could only access the type/module index spaces as they existed at the point of the module's declaration in the `module` section.  In the text format, would translate to only being able to access the types/modules textually preceding it.

Reply:
Hm ok interesting. I don't really grok the idea of a type export or import myself, but this does mean that a validator will have to basically save the state of the index space for each module declaration, since the module definition happens so much later in the module.

Reply:
To see a hypothetical example of the type of fun with type imports/exports, check out [this slide](https://docs.google.com/presentation/d/1Cz4yopK-V6PIq7JMJiRNUWlDlHLTnIvwhdetP5QBVL4/edit#slide=id.g874aa75317_0_303).  A cross-module cycle would look something like:
```wasm
(module
  (module $M
    (type $T (export "T") (func (param (ref $X))))
  )
  (instance $m (instantiate $M))
  (alias $X (type $m $T))
)
```
which is allowed by the forward reference of `$X`.  Fortunately, because index spaces are immutable, I think to save the state of the index space, you practically only have to snapshot the two lengths of the module and type index spaces at the point of each modules' declaration.